prometheus/metrics.go

```diff
@@ -459,52 +495,104 @@ func getAllRequestRates(api v1.API, namespace string, ratesInterval string) (mod
+	in, err := getRequestRatesForLabel(api, k8s, namespace, time.Now(), lbl, ratesInterval)
 	if err != nil {
 		return model.Vector{}, err
 	}
 	return in, nil
 }
 
-func getItemRequestRates(api v1.API, namespace, item, itemLabelSuffix, ratesInterval string) (model.Vector, model.Vector, error) {
+func getItemRequestRates(api v1.API, k8s kubernetes.IstioClientInterface, namespace, item, itemLabelSuffix, ratesInterval string) (model.Vector, model.Vector, error) {
 	lblIn := fmt.Sprintf(`reporter="destination",destination_workload_namespace="%s",destination_%s="%s"`, namespace, itemLabelSuffix, item)
 	outReporter := "source"
 	if config.Get().IstioNamespace == namespace {
 		outReporter = "destination"
 	}
 	lblOut := fmt.Sprintf(`reporter="%s",source_workload_namespace="%s",source_%s="%s"`, outReporter, namespace, itemLabelSuffix, item)
-	in, err := getRequestRatesForLabel(api, time.Now(), lblIn, ratesInterval)
+	in, err := getRequestRatesForLabel(api, k8s, namespace, time.Now(), lblIn, ratesInterval)
 	if err != nil {
 		return model.Vector{}, model.Vector{}, err
 	}
-	out, err := getRequestRatesForLabel(api, time.Now(), lblOut, ratesInterval)
+	out, err := getRequestRatesForLabel(api, k8s, namespace, time.Now(), lblOut, ratesInterval)
 	if err != nil {
 		return model.Vector{}, model.Vector{}, err
 	}
 	return in, out, nil
 }
 
-func getRequestRatesForLabel(api v1.API, time time.Time, labels, ratesInterval string) (model.Vector, error) {
+func getRequestRatesForLabel(api v1.API, k8s kubernetes.IstioClientInterface, namespace string, time time.Time, labels, ratesInterval string) (model.Vector, error) {
+	log.Debugf("[getRequestRatesForLabel] Requested rates interval for <%s> namespace: %v", namespace, ratesInterval)
+	ratesInterval, err := adjustRateInterval(k8s, namespace, time, ratesInterval)
+	if err != nil {
+		log.Debugf("[getRequestRatesForLabel] AdjustRateInterval error: %v", err)
+		return model.Vector{}, err
+	}
+	log.Debugf("[getRequestRatesForLabel] Rates interval for <%s> namespace, effective: %v", namespace, ratesInterval)
+
 	query := fmt.Sprintf("rate(istio_requests_total{%s}[%s])", labels, ratesInterval)
 	result, err := api.Query(context.Background(), query, time)
 	if err != nil {
 		return model.Vector{}, err
 	}
 	return result.(model.Vector), nil
 }
+
+func adjustRateInterval(k8s kubernetes.IstioClientInterface, namespace string, queryTime time.Time, ratesInterval string) (string, error) {
+	namespaceData, err := k8s.GetNamespace(namespace)
+	if err != nil {
+		return "", err
+	}
+
+	startTime, err := getStartForInterval(queryTime, ratesInterval)
+	if err != nil {
+		return "", err
+	}
+
+	if startTime.Before(namespaceData.CreationTimestamp.Time) {
+		interval := queryTime.Sub(namespaceData.CreationTimestamp.Time)
+		return fmt.Sprintf("%vs", int(interval.Seconds())), nil
+	}
+
+	return ratesInterval, nil
+}
+
+func getStartForInterval(baseTime time.Time, ratesInterval string) (time.Time, error) {
```

Comment:
I wonder if the prometheus go library already has a method to parse rate interval strings. @simonpasquier certainly knows?

Reply:
You're probably looking for [`github.com/prometheus/common/model.Duration`](https://godoc.org/github.com/prometheus/common/model#Duration).

Reply:
more specifically `ParseDuration`
https://github.com/prometheus/common/blob/master/model/time.go#L182